File: src-tauri/src/commands/credentials/desktop_bridges.rs
//! Tauri commands for desktop bridge execution and local agent runtime.

use std::sync::Arc;

use serde_json::Value;
use tauri::State;

use crate::engine::desktop_bridges::BridgeActionResult;
use crate::engine::desktop_runtime::{
    BridgeConfig, DesktopPlan, PlanExecutionResult, RuntimeStatus,
};
use crate::engine::desktop_security::{self, DesktopCapability};
use crate::error::AppError;
use crate::AppState;

// ── Single bridge action ─────────────────────────────────────────────

/// Execute a single action on a desktop bridge.
///
/// The caller provides the bridge name, action JSON, and config overrides.
/// Security capabilities are checked before execution.
#[tauri::command]
pub async fn execute_desktop_bridge(
    state: State<'_, Arc<AppState>>,
    bridge: String,
    action: Value,
    config: Option<BridgeConfig>,
) -> Result<BridgeActionResult, AppError> {
    // Map bridge name to connector for permission check
    let connector_name = format!("desktop_{}", bridge);
    let required_cap = match bridge.as_str() {
        "vscode" | "docker" | "terminal" => DesktopCapability::ProcessSpawn,
        "obsidian" => DesktopCapability::FileRead,
        _ => return Err(AppError::Validation(format!("Unknown bridge: {bridge}"))),
    };

    desktop_security::check_permission(
        &state.desktop_approvals,
        &connector_name,
        &required_cap,
    )?;

    let config = config.unwrap_or_default();

    match bridge.as_str() {
        "vscode" => {
            let action = serde_json::from_value(action)
                .map_err(|e| AppError::Validation(format!("Invalid VS Code action: {e}")))?;
            let binary = config.vscode_binary.as_deref().unwrap_or("code");
            crate::engine::desktop_bridges::vscode::execute(binary, action).await
        }
        "docker" => {
            let action = serde_json::from_value(action)
                .map_err(|e| AppError::Validation(format!("Invalid Docker action: {e}")))?;
            let binary = config.docker_binary.as_deref().unwrap_or("docker");
            crate::engine::desktop_bridges::docker::execute(binary, action).await
        }
        "terminal" => {
            let action = serde_json::from_value(action)
                .map_err(|e| AppError::Validation(format!("Invalid Terminal action: {e}")))?;
            let shell = config.terminal_shell.as_deref().unwrap_or("bash");
            crate::engine::desktop_bridges::terminal::execute(shell, action, &config.env_vars)
                .await
        }
        "obsidian" => {
            let action = serde_json::from_value(action)
                .map_err(|e| AppError::Validation(format!("Invalid Obsidian action: {e}")))?;
            let vault = config
                .obsidian_vault_path
                .as_deref()
                .ok_or_else(|| AppError::Validation("Obsidian vault path not configured".into()))?;
            crate::engine::desktop_bridges::obsidian::execute(
                vault,
                config.obsidian_api_port,
                config.obsidian_api_key.as_deref(),
                action,
            )
            .await
        }
        _ => Err(AppError::Validation(format!("Unknown bridge: {bridge}"))),
    }
}

// ── Plan execution (runtime) ─────────────────────────────────────────

/// Execute a multi-step desktop plan via the local agent runtime.
#[tauri::command]
pub async fn execute_desktop_plan(
    state: State<'_, Arc<AppState>>,
    plan: DesktopPlan,
    config: Option<BridgeConfig>,
) -> Result<PlanExecutionResult, AppError> {
    let config = config.unwrap_or_default();
    state
        .desktop_runtime
        .execute_plan(plan, &state.desktop_approvals, &config)
        .await
}

/// Get the current runtime status (active plan, progress).
#[tauri::command]
pub async fn get_desktop_runtime_status(
    state: State<'_, Arc<AppState>>,
) -> Result<RuntimeStatus, AppError> {
    Ok(state.desktop_runtime.get_status().await)
}

/// Get the cached result of a previously executed plan.
#[tauri::command]
pub async fn get_desktop_plan_result(
    state: State<'_, Arc<AppState>>,
    plan_id: String,
) -> Result<Option<PlanExecutionResult>, AppError> {
    Ok(state.desktop_runtime.get_result(&plan_id).await)
}

File: src/api/system/desktopBridges.ts
import { invokeWithTimeout as invoke } from "@/lib/tauriInvoke";

// -- Bridge action types ----------------------------------------------

export interface BridgeActionResult {
  success: boolean;
  output: string;
  error: string | null;
  duration_ms: number;
  bridge: string;
  action: string;
}

// VS Code actions
export type VsCodeAction =
  | { action: "OpenFile"; params: { path: string; line?: number } }
  | { action: "OpenFolder"; params: { path: string } }
  | { action: "DiffFiles"; params: { left: string; right: string } }
  | { action: "ListExtensions" }
  | { action: "InstallExtension"; params: { extension_id: string } }
  | { action: "RunTask"; params: { task_name: string; folder?: string } }
  | { action: "Version" };

// Docker actions
export type DockerAction =
  | { action: "ListContainers"; params: { all: boolean } }
  | { action: "ListImages" }
  | { action: "StartContainer"; params: { container: string } }
  | { action: "StopContainer"; params: { container: string } }
  | { action: "RestartContainer"; params: { container: string } }
  | { action: "ContainerLogs"; params: { container: string; tail?: number } }
  | { action: "InspectContainer"; params: { container: string } }
  | { action: "Exec"; params: { container: string; command: string[] } }
  | { action: "ComposeUp"; params: { file?: string; detach: boolean } }
  | { action: "ComposeDown"; params: { file?: string } }
  | { action: "ComposePs"; params: { file?: string } }
  | { action: "SystemInfo" }
  | { action: "Version" };

// Terminal actions
export type TerminalAction =
  | { action: "Execute"; params: { command: string[]; working_dir?: string } }
  | { action: "ReadFile"; params: { path: string } }
  | { action: "WriteFile"; params: { path: string; content: string } }
  | { action: "ListDir"; params: { path: string } }
  | { action: "PathExists"; params: { path: string } };

// Obsidian actions
export type ObsidianAction =
  | { action: "ListNotes"; params: { folder?: string } }
  | { action: "ReadNote"; params: { path: string } }
  | { action: "WriteNote"; params: { path: string; content: string } }
  | { action: "SearchNotes"; params: { query: string; max_results?: number } }
  | { action: "VaultStructure" }
  | { action: "AppendToNote"; params: { path: string; content: string } };

export type BridgeAction =
  | VsCodeAction
  | DockerAction
  | TerminalAction
  | ObsidianAction;

// -- Bridge config ----------------------------------------------------

export interface BridgeConfig {
  vscode_binary?: string;
  docker_binary?: string;
  terminal_shell?: string;
  obsidian_vault_path?: string;
  obsidian_api_port?: number;
  obsidian_api_key?: string;
  env_vars?: Record<string, string>;
}

// -- Plan types -------------------------------------------------------

export interface DesktopStep {
  id: string;
  bridge: string;
  action: BridgeAction;
  depends_on?: string;
  description: string;
}

export interface DesktopPlan {
  id: string;
  name: string;
  steps: DesktopStep[];
  created_at: string;
}

export interface StepResult {
  step_id: string;
  bridge_result: BridgeActionResult;
  injected_context?: string;
}

export interface PlanExecutionResult {
  plan_id: string;
  success: boolean;
  step_results: StepResult[];
  total_duration_ms: number;
  failed_step?: string;
}

export interface RuntimeStatus {
  active_plan: string | null;
  completed_steps: number;
  total_steps: number;
  current_step: string | null;
}

// -- Bridge metadata --------------------------------------------------

export type BridgeName = "vscode" | "docker" | "terminal" | "obsidian";

export const BRIDGE_INFO: Record<
  BridgeName,
  { label: string; description: string; connector: string }
> = {
  vscode: {
    label: "VS Code",
    description: "Open files, manage extensions, run tasks",
    connector: "desktop_vscode",
  },
  docker: {
    label: "Docker",
    description: "Manage containers, images, and compose stacks",
    connector: "desktop_docker",
  },
  terminal: {
    label: "Terminal",
    description: "Execute commands, read/write files",
    connector: "desktop_terminal",
  },
  obsidian: {
    label: "Obsidian",
    description: "Read, write, and search notes in your vault",
    connector: "desktop_obsidian",
  },
};

// -- API calls --------------------------------------------------------

/** Execute a single bridge action. */
export const executeDesktopBridge = (
  bridge: BridgeName,
  action: BridgeAction,
  config?: BridgeConfig,
) =>
  invoke<BridgeActionResult>("execute_desktop_bridge", {
    bridge,
    action,
    config: config ?? null,
  });

/** Execute a multi-step desktop plan. */
export const executeDesktopPlan = (
  plan: DesktopPlan,
  config?: BridgeConfig,
) =>
  invoke<PlanExecutionResult>("execute_desktop_plan", {
    plan,
    config: config ?? null,
  });

/** Get the current runtime execution status. */
export const getDesktopRuntimeStatus = () =>
  invoke<RuntimeStatus>("get_desktop_runtime_status", {});

/** Get the cached result of a previously executed plan. */
export const getDesktopPlanResult = (planId: string) =>
  invoke<PlanExecutionResult | null>("get_desktop_plan_result", { planId });
